### PR TITLE
Feature/popup

### PR DESCRIPTION
**Pull request by knegusen**
image
image
image

### CODE REVIEW

File: src/components/ui/popup/EnhancedPopper.tsx [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+  PopperComponent: typeof Popper;
+  placement?: Placement;
+  shadow?: boolean;
+  targetMinHeight?: number;
+  targetMinWidth?: number;
+}
+
+const TargetContainer = styled('div')<Partial<PopperProps>>(props => ({
+  backgroundColor: props.backgroundColor || '#fff',
+  boxShadow: props.shadow ? '0 0 8px 0 rgba(0, 0, 0, 0.15)' : undefined,
```
Comment (mattias800) [inline src/components/ui/popup/EnhancedPopper.tsx]: Should reuse styling from Shadow component. So we don't have to maintain several definitions of the same style.

File: src/components/ui/popup/EnhancedPopper.tsx [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+      PopperComponent,
+      placement: popupPlacement = 'auto-end',
+      targetMinWidth,
+      targetMinHeight,
+      shadow = true,
+    } = this.props;
+
+    return (
+      <PopperComponent placement={popupPlacement}>
+        {({ ref, style, placement, arrowProps }) => (
```
[the remark this answers is not in the document]
Comment (knegusen) [reply src/components/ui/popup/EnhancedPopper.tsx]: Handle click outside is already in this component so I make a private member method instead.

File: src/components/ui/popup/Popup.tsx [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+      referenceChildren,
+      shadow,
+      targetMinHeight,
+      targetMinWidth,
+    } = this.props;
+
+    return (
+      <Popper.Manager>
+        <Popper.Reference>
+          {({ ref }) => (
```
[the remark this answers is not in the document]
Comment (knegusen) [reply src/components/ui/popup/Popup.tsx]: Done

File: src/components/ui/popup/Popup.tsx [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+  onOpen: () => void;
+  Popper: PopperInterface;
+  placement?: Placement;
+  referenceChildren: React.ReactNode;
+  shadow?: boolean;
+  targetMinHeight?: number;
+  targetMinWidth?: number;
+}
+
+export class Popup extends React.Component<PopupProps> {
```
[the remark this answers is not in the document]
Comment (knegusen) [reply src/components/ui/popup/Popup.tsx]: Don't want to spend more time on this now so I leave it as a component

File: src/components/ui/popup/UncontrolledPopup.tsx [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  compose,
+  StateHandlerMap,
+  withProps,
+  withStateHandlers,
+} from 'recompose';
+import { Omit } from '../../../types';
+import { PopperInterface, Popup, PopupProps } from './Popup';
+
+export interface UncontrolledPopupProps {
+  backgroundColor?: Color;
```
[the remark this answers is not in the document]
Comment (knegusen) [reply src/components/ui/popup/UncontrolledPopup.tsx]: Sure!

File: src/components/ui/popup/EnhancedPopper.tsx [hunk 6]
```diff
@@ -0,0 +1,64 @@
+import { Placement } from 'popper.js';
+import * as React from 'react';
+import * as enhanceWithClickOutside from 'react-click-outside';
+import styled from 'react-emotion';
+import { Popper } from 'react-popper';
+
+export interface PopperProps {
+  backgroundColor?: string;
```
[the remark this answers is not in the document]
Comment (knegusen) [reply src/components/ui/popup/EnhancedPopper.tsx]: This one should not be used outside this package. JSDocs is written for Popup and UncontrolledPopup instead.

**Pull request merged**